Reopen log file when rotation is detected

ashie opened the pull request:
This is another solution for #106 to avoid using inotify.

> In the previous vesions, there was no way to detect log rotation event when log file was rotated.
> 
> It causes that rotated log file was grabbed by DaemonLogger. This was unexpected behavior.
> 
> This commit fixes by reopening log file correctly.
> 
> NOTE: This PR depends on inotify so only linux was supported.

Fix #106

kenhys commented:
I'll review this PR, just wait for a while.

kenhys reviewed (commented):
It may be useful to explain why extending behavior with a module and that intention a bit in the commit message.

kenhys commented on lib/serverengine/daemon_logger.rb:
```diff
@@ -55,6 +55,7 @@ def logdev=(logdev)
         old_file_dev = @file_dev
         @file_dev = LogDevice.new(logdev, shift_age: @rotate_age, shift_size: @rotate_size)
+        @file_dev.extend(RotationAware) if @rotate_age == 0 && !ServerEngine.windows?
         old_file_dev.close if old_file_dev
```
If log_rotate_age was not set, log_rotate_age: 5(default) is used.
Does it work intentionally in such a case?

ashie replied:
This change intend to reopen the log file only when the application doesn't rotate the file by itself.
(When such case, the application expects that the log is rotated by external tool.)

When `log_rotate_age` is non-zero (including `nil`), the log is rotated by the application.
Please see also the implementation of `LogDevice` of Ruby:

https://github.com/ruby/ruby/blob/abc828bc67ff42ddb42f5bd034cec185d6b8821d/lib/logger/log_device.rb#L14-L28
https://github.com/ruby/ruby/blob/abc828bc67ff42ddb42f5bd034cec185d6b8821d/lib/logger/log_device.rb#L123-L136

Fluentd sets `log_rotate_age` to `0` by default.

ashie replied:
> Fluentd sets `log_rotate_age` to `0` by default.

Sorry, it's wrong, it sets `nil` by default.

kenhys replied:
I've overlooked that situation, thanks.
So go forward this PR.

ashie replied:
Or it would be better that we always enable it, so that users can delete log file manually to recover free storage space.

ashie replied:
I'll remove the `if` statement.

ashie replied:
Removed in 88420811ac563225a5420371b295968f99f9cfa5

kenhys reviewed (approved):
I support this approach, could you solve the remaining concerns?, then it is OK for me.

ashie commented:
> It may be useful to explain why extending behavior with a module and that intention a bit in the commit message.

Added in 88420811ac563225a5420371b295968f99f9cfa5

ashie commented:
I've noticed that the current patch isn't enough.
`@ino` should be also updated when `reopen` is called directly by applications.

ashie commented:
> I've noticed that the current patch isn't enough. `@ino` should be also updated when `reopen` is called directly by applications.

Fixed in e2e8824ddd039adaea22f5f9ba71cecd5ba67523
